Iter13

skiphead opened the pull request.

skiphead committed:
Инкремент 10

Добавьте в сервис функциональность подключения к базе данных. В качестве СУБД используйте PostgreSQL не ниже 10 версии.
Добавьте в сервис хендлер GET /ping, который при запросе проверяет соединение с базой данных. При успешной проверке хендлер должен вернуть HTTP-статус 200 OK, при неуспешной — 500 Internal Server Error.
Строка с адресом подключения к БД должна получаться из переменной окружения DATABASE_DSN или флага командной строки -d.

skiphead committed:
Инкремент 10

fix test

skiphead committed:
Инкремент 10

fix pool

skiphead committed:
Инкремент 10

fix args

skiphead committed:
Инкремент 10

fix test

skiphead committed:
Инкремент 11

Добавлена поддержка бд
Миграции

skiphead committed:
Инкремент 11

fix test handler

skiphead committed:
Инкремент 11

fix param originalURL

skiphead committed:
Инкремент 11

fix param originalURL

skiphead committed:
Инкремент 11

fix param originalURL

skiphead committed:
Инкремент 12

Добавьте новый хендлер POST /api/shorten/batch

skiphead committed:
Инкремент 12

fix test )

skiphead committed:
Инкремент 12

fix struct field CorrelationId should be CorrelationID

skiphead committed:
Инкремент 13

skiphead committed:
Инкремент 13 fix unuse err

skiphead committed:
Инкремент 13 fix unuse err

skiphead committed:
Инкремент 13 fix unuse err

skiphead committed:
Инкремент 13 fix /n

skiphead committed:
Инкремент 13 fix gzip

skiphead committed:
Инкремент 13 refactor code

skiphead committed:
Инкремент 13 refactor
централизованная обработка ошибки конфликтов
убрана избыточная проверка сжатия тела

skiphead committed:
Инкремент 13 refactor
пакет log полностью заменен на zap

skiphead committed:
Инкремент 13 refactor main, added go doc and package description.

skiphead committed:
Инкремент 13 fix init pool

skiphead committed:
Инкремент 13 fix init pool

Perederey commented on internal/usecase/storage.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// IsDuplicateError унифицированная проверка на ошибку дублирования
+func (s *urlUseCase) IsDuplicateError(err error) error {
```
Метод IsDuplicateError странно работает - он возвращает error если это дубликат, а nil если не дубликат. По naming ожидается bool. Переделай на возврат bool

skiphead replied:
Поправил.

Perederey commented on internal/usecase/storage.go:
```diff
@@ -0,0 +1,186 @@
+	return shortURL, nil
+}
+
+// BatchSave сохраняет пакет URL и возвращает сокращенные версии
```
В BatchSave ты не обрабатываешь конфликты дублирования, хотя задание требует это для всех методов создания URL. Если в батче придет уже существующий URL, то будет ошибка, а не 409 статус. Нужно добавить такую же логику как в Save - проверять на дубликат и возвращать существующий short code

skiphead replied:
Если нам нужно обработать конфликт для этого есть IsDuplicateError эта функция принимает ошибку со значением "23505" то возвращаем true.

Perederey commented on pkg/utils/utils.go:
```diff
@@ -23,3 +24,16 @@ func IsValidURL(s string) bool {
+	keyLength     = 8
+	randomCharset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+)
+
```
Ошибка безопасности - используешь math/rand для генерации ключей URL. Это предсказуемо и небезопасно. Лучше использовать crypto/rand

Perederey commented on internal/domain/repository/postgres.go:
```diff
@@ -0,0 +1,392 @@
+	ErrNotFound = errors.New("not found")
+)
+
+type URLRepository interface {
```
URLRepository возвращает entity.ShortURL из domain слоя, но это нарушает чистую архитектуру - repository должен возвращать свои модели, а преобразование в domain entities делать в usecase. Сейчас у тебя repository знает о domain entities, что создает циклическую зависимость на концептуальном уровне

Perederey commented on internal/usecase/storage.go:
```diff
@@ -0,0 +1,186 @@
+func (s *urlUseCase) Save(ctx context.Context, originalURL string) (*entity.ShortURL, error) {
+	shortCode := utils.GenerateRandomKey()
+
+	if !s.isDatabaseAvailable(ctx) {
```
Намудрил тут немного с дублированием fallback логики - в каждом методе (Save, Get, GetByOriginalURL, BatchSave) повторяется одна и та же проверка isDatabaseAvailable. Вынеси это в отдельный метод

skiphead replied:
Тут все просто, если база не доступна работаем с файлом.

skiphead replied:
В текущей реализации isDatabaseAvailable уже вынесен и возвращает bool, и если база недоступна, мы используем файловое хранилище.

Perederey reviewed (commented):
Привет! Хорошо справился с заданием, круто что сделал разделение на слои (repository, usecase, handler) и добавил обработку конфликтов дублирования.

Из того, на что стоит обратить внимание:
- BatchSave вообще не обрабатывает конфликты дублирования, хотя задание требует это для всех методов создания
- GenerateRandomKey использует небезопасный math/rand вместо crypto/rand, это потенциальная уязвимость, это хоть и поправили в новых версиях, но все равно лучше использовать пока-что crypto
- В usecase дублируется fallback логика для каждого метода, лучше вынести в отдельный слой или использовать Chain of Responsibility

А так молодец! Продолжай в том же духе, архитектура в целом правильная!

skiphead committed:
Инкремент 13 refactor IsDuplicateError теперь возвращает тип bool. Если true то обнаружен дубликат.

skiphead committed:
Инкремент 13 refactor IsDuplicateError теперь возвращает тип bool. Если true то обнаружен дубликат.
Для генерации случайного ключа библиотека math/rand заменена на crypto/rand для улучшения безопасности.

skiphead committed:
Инкремент 13 refactor проверка дубликата убрана в репозиторий

skiphead committed:
Инкремент 13 refactor обработка дубликатов в Bach функции

skiphead committed:
Инкремент 13 refactor обработка дубликатов в Bach функции

skiphead committed:
Инкремент 13 refactor обработка дубликатов в Bach функции

skiphead committed:
Инкремент 13 refactor
